File: Sources/MoongateCore/SubtitleSourceDecisionEngine.swift
import Foundation

/// 统一的字幕源决策权威。把"质量评估(门每候选只跑一次) + 是否生成 Whisper 的计划 + 多候选择优(tie-break)"
/// 收敛到单一处，供下载后执行（`SubtitleSourceResolver` / `QueueManager`）与下载前 UI 预测
/// （`SubtitleSourceDecision`）共享同一套规则、阈值与裁决口径，消除"UI 说用平台字幕、实际却生成 Whisper"
/// 这类不一致，以及 QueueManager 里"门(布尔) + 评分器(5级)"双裁决叠加、门被跑两次的问题。
///
/// 反冲突契约（与 [[moongate-ux-asr-productization-review]] 一致）：
/// - **门 `gateUsable` 只答"这个候选自身是否可用"**，是"是否生成 Whisper"的唯一依据；它绝不看时序、
///   绝不跨源比较（铁律见 `PlatformSubtitleQualityGate`）。
/// - **`score`/`verdict` 只答"真实候选里谁更好"**，用于排名与 tie-break，不参与"是否生成"。
/// - 不变式：`gateUsable == false ⇒ verdict <= .lowConfidence`（门的致命扣分足够把分数压到 usable 以下）。
///
/// tie-break：平分时取**更可信来源**（`sourceKindRank` 更小，manual<auto<local）。已实测确认 Swift
/// `max(by:)` 与 C# `OrderByDescending.ThenBy` 在此口径下一致，两端不再各写一套。
public enum SubtitleSourceDecisionEngine {

    /// 单候选评估：对一个候选只跑一次门（经 `SubtitleQualityScorer` 内部），并暴露门的权威裁决与排名分数。
    public struct Assessment: Equatable, Sendable {
        public let candidateID: String
        public let kind: SubtitleSourceKind
        public let languageCode: String
        public let score: Double
        public let verdict: SubtitleQualityVerdict
        public let gateUsable: Bool
        public let gateReasons: [PlatformSubtitleQualityGate.Reason]
        public let reasons: [String]
        public let report: PlatformSubtitleQualityGate.SubtitleSourceQualityReport?
        public let hasFile: Bool
    }

    /// 执行端"拿到平台自动字幕后该怎么办"的计划。由 policy + 平台候选的门裁决唯一决定。
    public enum ASRPlan: Equatable, Sendable {
        /// 直接用当前平台/人工/导入源，无需生成。
        case none
        /// 生成 Whisper，再让引擎在平台源与本地源之间择优。
        case generateLocalASRThenChoose
        /// 平台不可用且无生成器：沿用平台源、记录原因给 UI 提示（不阻塞）。
        case keepPlatformRecordReasons([PlatformSubtitleQualityGate.Reason])
    }

    // MARK: - 阈值（与 fixture `subtitleSourceDecision` 段两端契约对齐）

    /// `.autoBest` 下，平台自动字幕 verdict 低于此档（即 lowConfidence/unusable）时也重生成 Whisper。
    /// 这是把原本散落在 QueueManager 的 `|| score <= .lowConfidence` 收敛为**具名、可测、单类型**的质量地板，
    /// 而非把"门布尔"与"评分枚举"混在一起 OR。门不可用(`!gateUsable`)始终触发，与本地板取并集。
    public static let autoBestRegenerateBelow: SubtitleQualityVerdict = .usable

    // MARK: - 评估

    public static func assess(
        candidate: SubtitleSourceCandidate,
        requestedSourceLanguageCode: String?,
        videoDurationSeconds: Double?
    ) -> Assessment {
        let score = SubtitleQualityScorer.score(
            candidate: candidate,
            requestedSourceLanguageCode: requestedSourceLanguageCode,
            videoDurationSeconds: videoDurationSeconds
        )
        return Assessment(
            candidateID: score.candidateID,
            kind: score.kind,
            languageCode: score.languageCode,
            score: score.score,
            verdict: score.verdict,
            gateUsable: score.gateUsable,
            gateReasons: score.gateReasons,
            reasons: score.reasons,
            report: score.report,
            hasFile: candidate.fileURL != nil
        )
    }

    // MARK: - 是否生成 Whisper

    /// 给定 policy 与平台自动字幕的门裁决，决定执行端动作。`platform` 为 nil 表示没有平台自动字幕候选。
    public static func generationPlan(
        policy: SubtitleSourcePolicy,
        platform: Assessment?,
        localASRAvailable: Bool,
        cloudASRAvailable: Bool
    ) -> ASRPlan {
        switch policy {
        case .forcePlatform, .preferPlatform, .cloudASR, .importedFile:
            return .none
        case .forceLocalASR, .compareLocalASR:
            return localASRAvailable ? .generateLocalASRThenChoose : noGenerator(platform)
        case .preferLocalASR:
            guard let platform else {
                return localASRAvailable ? .generateLocalASRThenChoose : .none
            }
            if platform.gateUsable { return .none }
            return localASRAvailable ? .generateLocalASRThenChoose : noGenerator(platform)
        case .autoBest:
            guard let platform else { return .none }
            let needsRegen = !platform.gateUsable || platform.verdict < autoBestRegenerateBelow
            guard needsRegen else { return .none }
            return localASRAvailable ? .generateLocalASRThenChoose : noGenerator(platform)
        }
    }

    private static func noGenerator(_ platform: Assessment?) -> ASRPlan {
        .keepPlatformRecordReasons(platform?.gateReasons ?? [])
    }

    // MARK: - 多候选择优（tie-break）

    /// 在已评估的候选中按 `score + policyBoost` 选最优；平分取更可信来源（更小 `sourceKindRank`）。
    /// 只在 `selectableIDs`（有文件的候选）里选。返回胜出候选的 id，无可选时返回 nil。
    public static func choose(
        policy: SubtitleSourcePolicy,
        assessments: [Assessment],
        selectableIDs: Set<String>
    ) -> String? {
        let selectable = assessments.filter { selectableIDs.contains($0.candidateID) }
        guard !selectable.isEmpty else { return nil }
        return selectable.max(by: { lhs, rhs in
            let lhsRank = lhs.score + policyBoost(lhs.kind, policy)
            let rhsRank = rhs.score + policyBoost(rhs.kind, policy)
            if lhsRank != rhsRank { return lhsRank < rhsRank }
            // 平分：更可信来源(更小 sourceKindRank)胜出。`max(by:)` 取"非更小"者，故此处返回
            // "lhs 更不可信(rank 更大)即视为更小" → 最终选出 rank 最小者。
            return sourceKindRank(lhs.kind) > sourceKindRank(rhs.kind)
        })?.candidateID
    }

    // MARK: - 排名权重（fixture 契约：subtitleSourceDecision.policyBoost / sourceKindRank）

    static func policyBoost(_ kind: SubtitleSourceKind, _ policy: SubtitleSourcePolicy) -> Double {
        switch policy {
        case .autoBest:
            return 0
        case .preferPlatform:
            return isPlatform(kind) ? 12 : 0
        case .forcePlatform:
            return isPlatform(kind) ? 10_000 : -10_000
        case .preferLocalASR:
            return kind == .localASR ? 12 : 0
        case .forceLocalASR:
            return kind == .localASR ? 10_000 : -10_000
        case .compareLocalASR:
            return 0
        case .cloudASR:
            return kind == .cloudASR ? 10_000 : -10_000
        case .importedFile:
            return kind == .importedFile ? 10_000 : -10_000
        }
    }

    static func isPlatform(_ kind: SubtitleSourceKind) -> Bool {
        kind == .manual || kind == .platformAuto || kind == .hlsManifest
    }

    static func sourceKindRank(_ kind: SubtitleSourceKind) -> Int {
        switch kind {
        case .manual: return 0
        case .importedFile: return 1
        case .hlsManifest: return 2
        case .platformAuto: return 3
        case .cloudASR: return 4
        case .localASR: return 5
        }
    }
}

File: windows/MoongateCore/SubtitleSourceDecisionEngine.cs
namespace Moongate.Core;

/// <summary>
/// 统一的字幕源决策权威（C# 镜像 Swift `SubtitleSourceDecisionEngine`）。把"质量评估(门每候选只跑一次) +
/// 是否生成 Whisper 的计划 + 多候选择优(tie-break)"收敛到单一处，供下载后执行（SubtitleSourceResolver /
/// Queue）共享同一套规则，消除"门(布尔) || 评分器(枚举)"双裁决叠加与门被多次跑的问题。
///
/// 反冲突契约：门 <c>GateUsable</c> 只答"这个候选自身是否可用"，是"是否生成 Whisper"的唯一依据，绝不看时序、
/// 绝不跨源比较；<c>Score</c>/<c>Verdict</c> 只答"真实候选里谁更好"，用于排名与 tie-break。
/// </summary>
public static class SubtitleSourceDecisionEngine
{
    /// <summary>单候选评估：对一个候选只跑一次门（经 SubtitleQualityScorer 内部），暴露门权威裁决与排名分数。</summary>
    public sealed record Assessment(
        string CandidateId,
        SubtitleSourceKind Kind,
        string LanguageCode,
        double Score,
        SubtitleQualityVerdict Verdict,
        bool GateUsable,
        IReadOnlyList<PlatformSubtitleQualityGate.Reason> GateReasons,
        IReadOnlyList<string> Reasons,
        PlatformSubtitleQualityGate.SubtitleSourceQualityReport? Report,
        bool HasFile);

    public enum AsrPlanKind
    {
        /// <summary>直接用当前平台/人工/导入源，无需生成。</summary>
        None,
        /// <summary>生成 Whisper，再让引擎在平台源与本地源之间择优。</summary>
        GenerateLocalAsrThenChoose,
        /// <summary>平台不可用且无生成器：沿用平台源、记录原因给 UI 提示（不阻塞）。</summary>
        KeepPlatformRecordReasons,
    }

    public sealed record AsrPlan(AsrPlanKind Kind, IReadOnlyList<PlatformSubtitleQualityGate.Reason> Reasons);

    /// <summary>
    /// `.autoBest` 下，平台自动字幕 verdict 低于此档（即 LowConfidence/Unusable）时也重生成。这是把原本散落在
    /// Queue 的 <c>|| platformScore.Verdict &lt;= LowConfidence</c> 收敛为具名、可测、单类型的质量地板。
    /// </summary>
    public const SubtitleQualityVerdict AutoBestRegenerateBelow = SubtitleQualityVerdict.Usable;

    public static Assessment Assess(
        SubtitleSourceCandidate candidate,
        string? requestedSourceLanguageCode,
        double? videoDurationSeconds)
    {
        var score = SubtitleQualityScorer.Score(candidate, requestedSourceLanguageCode, videoDurationSeconds);
        return new Assessment(
            score.CandidateId,
            score.Kind,
            score.LanguageCode,
            score.Score,
            score.Verdict,
            score.GateUsable,
            score.GateReasons ?? [],
            score.Reasons,
            score.Report,
            !string.IsNullOrWhiteSpace(candidate.FilePath) && File.Exists(candidate.FilePath));
    }

    public static AsrPlan GenerationPlan(
        SubtitleSourcePolicy policy,
        Assessment? platform,
        bool localAsrAvailable,
        bool cloudAsrAvailable)
    {
        switch (policy)
        {
            case SubtitleSourcePolicy.ForcePlatform:
            case SubtitleSourcePolicy.PreferPlatform:
            case SubtitleSourcePolicy.CloudAsr:
            case SubtitleSourcePolicy.ImportedFile:
                return new AsrPlan(AsrPlanKind.None, []);
            case SubtitleSourcePolicy.ForceLocalAsr:
            case SubtitleSourcePolicy.CompareLocalAsr:
                return localAsrAvailable
                    ? new AsrPlan(AsrPlanKind.GenerateLocalAsrThenChoose, [])
                    : NoGenerator(platform);
            case SubtitleSourcePolicy.PreferLocalAsr:
                if (platform is null)
                {
                    return localAsrAvailable
                        ? new AsrPlan(AsrPlanKind.GenerateLocalAsrThenChoose, [])
                        : new AsrPlan(AsrPlanKind.None, []);
                }
                if (platform.GateUsable) return new AsrPlan(AsrPlanKind.None, []);
                return localAsrAvailable
                    ? new AsrPlan(AsrPlanKind.GenerateLocalAsrThenChoose, [])
                    : NoGenerator(platform);
            case SubtitleSourcePolicy.AutoBest:
            default:
                if (platform is null) return new AsrPlan(AsrPlanKind.None, []);
                var needsRegen = !platform.GateUsable || platform.Verdict < AutoBestRegenerateBelow;
                if (!needsRegen) return new AsrPlan(AsrPlanKind.None, []);
                return localAsrAvailable
                    ? new AsrPlan(AsrPlanKind.GenerateLocalAsrThenChoose, [])
                    : NoGenerator(platform);
        }
    }

    private static AsrPlan NoGenerator(Assessment? platform) =>
        new(AsrPlanKind.KeepPlatformRecordReasons, platform?.GateReasons ?? []);

    /// <summary>
    /// 在已评估候选中按 Score + PolicyBoost 选最优；平分取更可信来源（更小 SourceKindRank）。
    /// 只在 selectableIds（有文件的候选）里选。无可选时返回 null。
    /// </summary>
    public static string? Choose(
        SubtitleSourcePolicy policy,
        IReadOnlyList<Assessment> assessments,
        ISet<string> selectableIds)
    {
        var selectable = assessments.Where(a => selectableIds.Contains(a.CandidateId)).ToArray();
        if (selectable.Length == 0) return null;
        return selectable
            .OrderByDescending(a => a.Score + PolicyBoost(a.Kind, policy))
            .ThenBy(a => SourceKindRank(a.Kind))
            .First()
            .CandidateId;
    }

    internal static double PolicyBoost(SubtitleSourceKind kind, SubtitleSourcePolicy policy) => policy switch
    {
        SubtitleSourcePolicy.AutoBest => 0,
        SubtitleSourcePolicy.PreferPlatform => IsPlatform(kind) ? 12 : 0,
        SubtitleSourcePolicy.ForcePlatform => IsPlatform(kind) ? 10_000 : -10_000,
        SubtitleSourcePolicy.PreferLocalAsr => kind == SubtitleSourceKind.LocalAsr ? 12 : 0,
        SubtitleSourcePolicy.ForceLocalAsr => kind == SubtitleSourceKind.LocalAsr ? 10_000 : -10_000,
        SubtitleSourcePolicy.CompareLocalAsr => 0,
        SubtitleSourcePolicy.CloudAsr => kind == SubtitleSourceKind.CloudAsr ? 10_000 : -10_000,
        SubtitleSourcePolicy.ImportedFile => kind == SubtitleSourceKind.ImportedFile ? 10_000 : -10_000,
        _ => 0,
    };

    internal static bool IsPlatform(SubtitleSourceKind kind) =>
        kind is SubtitleSourceKind.Manual or SubtitleSourceKind.PlatformAuto or SubtitleSourceKind.HlsManifest;

    internal static int SourceKindRank(SubtitleSourceKind kind) => kind switch
    {
        SubtitleSourceKind.Manual => 0,
        SubtitleSourceKind.ImportedFile => 1,
        SubtitleSourceKind.HlsManifest => 2,
        SubtitleSourceKind.PlatformAuto => 3,
        SubtitleSourceKind.CloudAsr => 4,
        SubtitleSourceKind.LocalAsr => 5,
        _ => 6,
    };
}
